File: android/core/src/main/kotlin/org/conduit/core/wire/ScreenPacking.kt
package org.conduit.core.wire

import java.io.ByteArrayOutputStream

/** Encoded video frame: parameter sets travel with every keyframe. */
data class EncodedVideoFrame(val isKeyframe: Boolean, val parameterSets: List<ByteArray>, val sampleData: ByteArray)

/** Packs/unpacks the ScreenFrame.data blob, matching Swift + Go:
 *  paramCount u8 | [len u32be | bytes]... | sampleData. */
object ScreenPacking {
    private const val MAX_PARAM_SETS = 8

    fun pack(f: EncodedVideoFrame): ByteArray = ByteArrayOutputStream().apply {
        val count = minOf(f.parameterSets.size, MAX_PARAM_SETS)
        write(count)
        for (i in 0 until count) {
            val set = f.parameterSets[i]
            write(set.size ushr 24 and 0xFF); write(set.size ushr 16 and 0xFF)
            write(set.size ushr 8 and 0xFF); write(set.size and 0xFF)
            write(set)
        }
        write(f.sampleData)
    }.toByteArray()

    fun unpack(data: ByteArray, isKeyframe: Boolean): EncodedVideoFrame {
        require(data.isNotEmpty()) { "truncated packed screen frame" }
        val count = data[0].toInt() and 0xFF
        require(count <= MAX_PARAM_SETS) { "too many parameter sets" }
        var cursor = 1
        val sets = ArrayList<ByteArray>(count)
        repeat(count) {
            require(cursor + 4 <= data.size) { "truncated" }
            var len = 0
            for (i in 0 until 4) { len = len shl 8 or (data[cursor + i].toInt() and 0xFF) }
            cursor += 4
            require(cursor + len <= data.size) { "truncated" }
            sets.add(data.copyOfRange(cursor, cursor + len))
            cursor += len
        }
        return EncodedVideoFrame(isKeyframe, sets, data.copyOfRange(cursor, data.size))
    }
}

File: core/wire/screenpacking.go
package wire

import (
	"encoding/binary"
	"errors"
)

// EncodedVideoFrame mirrors the Swift type: codec parameter sets travel with
// every keyframe; the sample data is the elementary stream.
type EncodedVideoFrame struct {
	IsKeyframe    bool
	ParameterSets [][]byte
	SampleData    []byte
}

const maxParameterSets = 8

var (
	ErrScreenTruncated  = errors.New("truncated packed screen frame")
	ErrTooManyParamSets = errors.New("too many parameter sets")
)

// PackScreenFrame serializes a frame into the ScreenFrame.Data blob:
// paramCount u8 | [len u32be | bytes]... | sampleData.
func PackScreenFrame(f EncodedVideoFrame) []byte {
	var out []byte
	count := len(f.ParameterSets)
	if count > maxParameterSets {
		count = maxParameterSets
	}
	out = append(out, byte(count))
	for i := 0; i < count; i++ {
		var lenBuf [4]byte
		binary.BigEndian.PutUint32(lenBuf[:], uint32(len(f.ParameterSets[i])))
		out = append(out, lenBuf[:]...)
		out = append(out, f.ParameterSets[i]...)
	}
	out = append(out, f.SampleData...)
	return out
}

// UnpackScreenFrame parses the ScreenFrame.Data blob.
func UnpackScreenFrame(data []byte, isKeyframe bool) (EncodedVideoFrame, error) {
	if len(data) < 1 {
		return EncodedVideoFrame{}, ErrScreenTruncated
	}
	count := int(data[0])
	if count > maxParameterSets {
		return EncodedVideoFrame{}, ErrTooManyParamSets
	}
	cursor := 1
	sets := make([][]byte, 0, count)
	for i := 0; i < count; i++ {
		if cursor+4 > len(data) {
			return EncodedVideoFrame{}, ErrScreenTruncated
		}
		l := int(binary.BigEndian.Uint32(data[cursor : cursor+4]))
		cursor += 4
		if cursor+l > len(data) {
			return EncodedVideoFrame{}, ErrScreenTruncated
		}
		sets = append(sets, append([]byte(nil), data[cursor:cursor+l]...))
		cursor += l
	}
	return EncodedVideoFrame{
		IsKeyframe:    isKeyframe,
		ParameterSets: sets,
		SampleData:    append([]byte(nil), data[cursor:]...),
	}, nil
}
